### PR TITLE
add int32 support for param seq_length of op reverse_sequence

### PR DESCRIPTION
**Pull request by suiyuan2009**
#3528

### CODE REVIEW

**Comment by ebrevdo** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/reverse_sequence_op.cc [hunk 1]
```diff
@@ -76,8 +76,8 @@ void CheckErrors(OpKernelContext* context, int batch_dim, int seq_dim) {
   }
 }
 
-template <>
-void CheckErrors<GPUDevice>(OpKernelContext* context, int batch_dim,
+template <typename Tlen>
```
Comment (ebrevdo) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: why do you need to template CheckErrors on Tlen?  I don't see you using Tlen anywhere.

Comment (suiyuan2009) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: I'll remove it. I see Jenkins output, and I'll modify gpu version. What does //tensorflow/core/ops/compat:backwards_compatibility_test do?

**Comment by suiyuan2009** [comment]
what does `Forward declarations of the functor specializations for GPU.` mean ? so many code for gpu specializations.

File: tensorflow/core/kernels/reverse_sequence_op.cc (hunk 1 above)
Comment (suiyuan2009) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: cpu version of CheckErrors need Tlen

File: tensorflow/core/kernels/reverse_sequence_op.cc [hunk 2]
```diff
@@ -97,7 +96,19 @@ void CheckErrors<GPUDevice>(OpKernelContext* context, int batch_dim,
...
+template <>
+void CheckErrors<GPUDevice, int32>(OpKernelContext* context, int batch_dim,
+                            int seq_dim) {
+  CheckErrorsGPU(context, batch_dim, seq_dim);
+}
+
+template <>
+void CheckErrors<GPUDevice, int64>(OpKernelContext* context, int batch_dim,
+                            int seq_dim) {
+  CheckErrorsGPU(context, batch_dim, seq_dim);
```
Comment (ebrevdo) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: i still don't understand why you can't just do:

template <typename Tlen>
void CheckErrors<GPUDevice, Tlen>(...) { check the errors }

?

Comment (suiyuan2009) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: that's the point, c++11 don't support partial specialization of function templates

Comment (rmlarsen) [inline tensorflow/core/kernels/reverse_sequence_op.cc]: Only full specialization is allowed for function templates in C++11. Partial specialization is only allowed for class templates.
http://en.cppreference.com/w/cpp/language/partial_specialization

**Comment by ebrevdo** [comment]
Jenkins, test this please.

**Comment by suiyuan2009** [comment]
@ebrevdo `//tensorflow/core/ops/compat:backwards_compatibility_test` this test failed, what does it mean?

**Comment by suiyuan2009** [comment]
@ebrevdo backwards_compatibility is fixed now.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Pull request merged**